Delegate the creation of the LBClient in the RibbonClient to a LBClientF...

bstick12 opened the pull request:
Delegate the creation of the LBClient in the RibbonClient to a LBClientFactory

cloudbees-pull-request-builder commented:
[NetflixOSS » feign » feign-pull-requests #59](https://netflixoss.ci.cloudbees.com/job/NetflixOSS/job/feign/job/feign-pull-requests/59/) SUCCESS
This pull request looks good

adriancole commented:
Few notes (from mobile)

 * I forgot that LBClient was a type we invented. Perhaps it is best to have the factory return what we actually need (abstractloadbalancer..) vs exposing the LBClient type. For example, you can make a caching decorator on top of the default factory without needing to know the concrete result type. In doing so, we expose the least possible interface. If there are other reasons you need LB client exposed then that's another story ;)

* RibbonClient is starting to look like it needs a builder. Instead of adding new public ctors, we probably want to make a nested Builder class at this point. There are similar classes you could look at for style reasons.

* nit.. it seems your changes reorder imports. Are you using the google style formatter?

adriancole commented:
Oh ic why you need to expose LBClient. Scratch bullet 1 above and replace with this. Make LBClient final and expose a static factory method instead of exposing the ctor. In general, ctors limit us. Ex we cant easily migrate to AutoValue when folks use ctors directly or can subtype our classes.

Also it seems clear enough to nest LBClientFactory as LBClient.Factory

Finally, I noticed exposing a getClient() method. Prefer client() as that's more conventional to feign.

cloudbees-pull-request-builder commented:
[NetflixOSS » feign » feign-pull-requests #62](https://netflixoss.ci.cloudbees.com/job/NetflixOSS/job/feign/job/feign-pull-requests/62/) SUCCESS
This pull request looks good

cloudbees-pull-request-builder commented:
[NetflixOSS » feign » feign-pull-requests #63](https://netflixoss.ci.cloudbees.com/job/NetflixOSS/job/feign/job/feign-pull-requests/63/) SUCCESS
This pull request looks good

bstick12 commented:
I've added the Builder for the RibbonClient and marked the old ctors as deprecated to maintain backwards compatibility.

I generally have Organize Imports turned on in Eclipse. I've reverted the imports that were re-organised but I don't think there is a consistent ordering across the source files.

adriancole commented on ribbon/src/main/java/feign/ribbon/RibbonClient.java:
```diff
@@ -78,4 +94,25 @@ public void loadDefaultValues() {
 
   }
   
+  public static class Builder {
```
nit final

adriancole commented on ribbon/src/main/java/feign/ribbon/LBClient.java:
```diff
@@ -35,19 +35,17 @@
   private final IClientConfig clientConfig;
 
-  LBClient(Client delegate, ILoadBalancer lb, IClientConfig clientConfig) {
+  public LBClient(ILoadBalancer lb, IClientConfig clientConfig) {
```
hide this and expose factory method

adriancole commented on ribbon/src/main/java/feign/ribbon/LBClient.java:
```diff
@@ -99,8 +99,12 @@ Request toRequest() {
           .request();
     }
 
+    Client client() {
```
ps why was this change made? (from field on LBClient to field RibbonRequest) I may have lost it during the review.

bstick12 replied:
It was moved here as otherwise the LBClientFactory would have the take the delegate and the clientname and I wanted to be able to key the LBClient just off the clientName.

adriancole replied:
ahh gotcha. thanks for the explanation!

adriancole commented on ribbon/src/main/java/feign/ribbon/LBClientFactory.java:
```diff
@@ -0,0 +1,25 @@
+
+  /**
+   * Uses {@link ClientFactory} static factories from ribbon to create an LBClient.
+   * */
```
nit bad format

adriancole commented on ribbon/src/main/java/feign/ribbon/LBClientFactory.java:
```diff
@@ -0,0 +1,25 @@
+import com.netflix.loadbalancer.ILoadBalancer;
+
+/**
+ * Factory to be used for the creation of {@link LBClient}'s for usage in {@link RibbonClient}
```
I appreciate the polish, but I'm not sure this clarifies much. Let's remove javadoc that don't pull their weight.

adriancole commented on ribbon/src/main/java/feign/ribbon/RibbonClient.java:
```diff
@@ -4,36 +4,53 @@
+  }
+
+  /**
+   * @deprecated Use the {@link RibbonClient#builder()}
```
`RibbonClient.builder().build()` is onerous for the simple case. how about just `RibbonClient.create()`?

```java
 public static RibbonClient create() {
   return new RibbonClient.builder().build();
 }
```

bstick12 replied:
agreed

adriancole commented:
couple last notes. nearly there! thanks for the follow-up.

wrt import order, I've used the google style format for idea and reformatted everything. I also do so on each change, and order tends to be deterministic. That said, I probably missed some files..

https://code.google.com/p/google-styleguide/source/browse/trunk/intellij-java-google-style.xml

cloudbees-pull-request-builder commented:
[NetflixOSS » feign » feign-pull-requests #64](https://netflixoss.ci.cloudbees.com/job/NetflixOSS/job/feign/job/feign-pull-requests/64/) SUCCESS
This pull request looks good

bstick12 commented:
Made the suggested changes. 

The Eclipse formatter configuration doesn't control the organization of imports its done as part of a different config section.

adriancole commented:
eek.. just looked at the idea import settings and noticed they are very
specific, so probably easy to get wrong manually. If you find a solution to
this problem for eclipse, please share. In the mean time, I'll try to keep
this difficulty in mind when reviewing contribs from eclipse IDE.

adriancole commented:
good work. so I am going to reword the commit while cherry-picking in order to save you from another nit change. will update in a few.

adriancole commented:
closed via 222793057764d138399fff8fc857a1649ef8705a (in master) and 4f6f8d4ecd9fe6ecb9ef0abe9eb96bb02aacaf6e in (7.x)
